fill empty cycle

Ashivu opened the pull request:
a new function "charge_discharge" is created to simulate cycles of filling and emptying of silo.

Ashivu committed:
charge_discharge in params and in void_migration.py

Ashivu committed:
Changes to include nu_g; added a chunk in move_voids - different for nu < nu_g and nu > nu_g

added a chunk in move_voids - different for nu < nu_g and nu > nu_g

Ashivu committed:
fill empty cycle

a new function "charge_discharge" is created to simulate cycles of filling and emptying of silo.

benjym commented on void_migration/plotter.py:
```diff
@@ -103,20 +103,31 @@ def plot_s_bar(y, s_bar, nu_time, p):
     plt.figure(fig)
     plt.clf()
+    ##################### if charge_discharge and mono-disperse #####################################
+    colors = [(1, 0, 0), (0, 0, 1)]
```
can you move this up to be defined with the rest of the colormaps?

benjym commented on void_migration/plotter.py:
```diff
@@ -103,20 +103,31 @@ def plot_s_bar(y, s_bar, nu_time, p):
         s_plot = np.nanmean(s, axis=2).T
     s_plot = np.ma.masked_where(np.isnan(s_plot), s_plot)
-    plt.pcolormesh(x, y, s_plot, cmap=orange_blue_cmap, vmin=0.001, vmax=0.01)
+    # plt.pcolormesh(x, y, s_plot, cmap=orange_blue_cmap, vmin=p.s_m, vmax=p.s_M)
```
this change will break all other use cases of this function. can you either make it a new command, or even better move it into the existing `plot_c` function, and have an optional function argument to choose between using the `c` variable or a fixed number of `s` increments?

benjym commented on void_migration/plotter.py:
```diff
@@ -103,20 +103,31 @@ def plot_s_bar(y, s_bar, nu_time, p):
     # plt.colorbar()
+    if p.charge_discharge:
+        # plt.pcolormesh(x, y, s_plot, cmap=orange_blue_cmap, vmin=p.s_m, vmax=p.s_M)
+        plt.pcolormesh(x, y, s_plot, cmap=cmap, vmin=0.000045, vmax=0.00004505)
```
we cant have hardcoded values like this here. can you calculate them from the input parameter file instead?

benjym commented on void_migration/plotter.py:
```diff
@@ -103,20 +103,31 @@ def plot_s_bar(y, s_bar, nu_time, p):
     plt.ylim(y[0], y[-1])
+    ticks = np.linspace(p.s_m, p.s_M, 3, endpoint=True)
     plt.subplots_adjust(left=0, right=1, bottom=0, top=1)
-    plt.colorbar(shrink=0.8,location='top',pad = 0.01)
```
can you make these colorbars optional?

benjym commented on void_migration/void_migration.py:
```diff
@@ -248,7 +250,7 @@ def move_voids(
                 # if perf_plate and i+1==perf_pts[1]: P_r *= perf_rate
 
-                P_tot = P_u + P_l + P_r
+                P_tot = (0.98*P_u) + (0.98*P_l) + (0.98*P_r)
```
what is this for? if you want more stability lower the stability cutoff in the latest version of void_migration

benjym commented on void_migration/void_migration.py:
```diff
@@ -439,11 +571,72 @@ def time_march(p):
+        # Time in 
+        T_in = [int((Mass_in[0]/Mass_in_per_u_t)*p.dt),int((Mass_in[1]/Mass_in_per_u_t)*p.dt),int((Mass_in[2]/Mass_in_per_u_t)*p.dt)]
+
+        # Time settle 
+        # T_settle = [T_in[0] + 3,T_in[1] + 3,T_in[2] + 3]
+        T_settle = [3, 3, 3]
+
+        # Time out
+        # T_out = [int((Mass_out[0]/Mass_out_per_u_t)*p.dt) + T_settle[0],int((Mass_out[1]/Mass_out_per_u_t)*p.dt) + T_settle[1],int((Mass_out[2]/Mass_out_per_u_t)*p.dt) + T_settle[2]]
+        T_out = [int((Mass_out[0]/Mass_out_per_u_t)*p.dt), int((Mass_out[1]/Mass_out_per_u_t)*p.dt), int((Mass_out[2]/Mass_out_per_u_t)*p.dt)]
+
+        print(T_in,T_settle,T_out)
+
+        tmp = 0
+        for i in range(no_of_cycles):
+
+            T_in[i] = T_in[i] + tmp
+            T_settle[i] = T_settle[i] + T_in[i]
+            T_out[i] = T_out[i] + T_settle[i]
+            tmp = T_out[i]
+
+        all_Ts = np.transpose([T_in,T_settle,T_out])   ## to arange the times in a cycle fashion
+
+        # Initialize dictionary
+        times = ["t_fill","t_settle","t_empty"]
+        op_arr = []
+        for i in range(no_of_cycles):
+            op_arr.append({x: {} for x in times})    ## creating empty dictionary with empty times
+
+        for j in range(no_of_cycles):
+            op_arr[j] = {key: all_Ts[j][i] for i, key in enumerate(op_arr[j])}   ##assigned the values for times t_fill t_settle and t_empty
+
+    print("EEEEEEEEEEEEEE",op_arr)
+
+
+    if p.charge_discharge:
+        p.nt = int(np.ceil(op_arr[-1].get('t_empty') / p.dt))
+
+    else:
+        p.nt = int(np.ceil(p.t_f / p.dt))
```
can you pull all of this out into a separate function and add a docstring?

benjym commented on void_migration/void_migration.py:
```diff
@@ -546,6 +765,9 @@ def time_march(p):
     plotter.plot_s_bar(y, s_bar_time, nu_time, p)
     plotter.plot_u_time(y, u_time, nu_time, p)
     np.save(p.folderName + "nu_t_x.npy", nu_time_x)
+    np.save(p.folderName + "mass_in_fun.npy", mass_in_fun)
+    np.save(p.folderName + "mass_count.npy", p_count)
+    np.save(p.folderName + "nu_non_zero_avg.npy", non_zero_nu_time)
```
the way saving is controlled has been updated in the new version. can you move this across?

benjym commented on void_migration/void_migration.py:
```diff
@@ -501,6 +698,19 @@ def time_march(p):
+            mass_in_fun.append(np.sum(mass_cal(den,s,p)))
+
+            if np.max(den) > 1.0:
+                print("*******************",np.max(den))
```
can you remove all of these diagnostics from the main output? if you need these to be run some times, pull this out of the main function into a separate function, and add an optional `debug` flag that you can set from the params file.

Ashivu committed:
Merge branch 'main' into SA

Ashivu committed:
Fill empty cycles

Made changes according to the comments on the previous pull request.

benjym commented on void_migration/void_migration.py:
```diff
@@ -466,6 +516,156 @@ def close_voids(u, v, s):
+
+                for k in range(p.nm):
+                    if ~np.isnan(s[i,j,k]):
+                        mass_k.append((4/3)*np.pi*(s[i,j,k]/2)**3*3950)  ##  volume of sphere s[i,j,k] x density of alumina particle x solid fraction nu[i,j]
```
why is the volume of the spheres here? every cell has the same volume?

benjym commented on void_migration/void_migration.py:
```diff
@@ -466,6 +516,156 @@ def close_voids(u, v, s):
+    op_arr : holds the values for t_fill, t_settle and t_empty for each cycle and for a desired number of cycles  
+
+    '''
+    if p.gsd_mode == 'mono':
```
again, not sure why you would need to know the volume of individual spheres?

benjym commented on void_migration/void_migration.py:
```diff
@@ -466,6 +516,156 @@ def close_voids(u, v, s):
+    rho_p = 3950    ## kg/m^3
+
+    # Total mass
+    M_total = p.nx * p.ny * p.nm * v_sphere * rho_p * p.nu_cs
```
is this the total mass in the system? if so it is incorrect. the total mass is just the number of cells filled with material (i.e. non-nan) times dx times dy times the density (in kg/m^2?). If you want to define the density in kg/m^3 then we need a nominal thickness of the cells.

benjym commented on void_migration/void_migration.py:
```diff
@@ -466,6 +516,156 @@ def close_voids(u, v, s):
+    # Total mass
+    M_total = p.nx * p.ny * p.nm * v_sphere * rho_p * p.nu_cs
+
+    free_par = 0.17             ## do not know why this. How to decide this?
```
what is this? whatever it is it shouldnt be hardcoded inside the function

benjym commented on void_migration/void_migration.py:
```diff
@@ -466,6 +516,156 @@ def close_voids(u, v, s):
+    else:
+        v_sphere = (4/3) * np.pi * ((p.s_m + p.s_M)/2/2)**3   ## m^3
+
+    rho_p = 3950    ## kg/m^3
```
this should be in the parameters file

benjym commented on void_migration/void_migration.py:
```diff
@@ -466,6 +516,156 @@ def close_voids(u, v, s):
+    T_in = [int((Mass_in[0]/Mass_in_per_u_t)*p.dt),int((Mass_in[1]/Mass_in_per_u_t)*p.dt),int((Mass_in[2]/Mass_in_per_u_t)*p.dt)]
+
+    # Time settle 
+    T_settle = [3, 3, 3]
```
this should be in the parameters file or at least a default value in defaults.json5

benjym commented on void_migration/void_migration.py:
```diff
@@ -466,6 +516,156 @@ def close_voids(u, v, s):
+    # Time out
+    T_out = [int((Mass_out[0]/Mass_out_per_u_t)*p.dt), int((Mass_out[1]/Mass_out_per_u_t)*p.dt), int((Mass_out[2]/Mass_out_per_u_t)*p.dt)]
+
+    print(T_in,T_settle,T_out)
```
any print statements should only run if a debug or verbose flag is passed

benjym commented on void_migration/void_migration.py:
```diff
@@ -507,7 +710,17 @@ def time_march(p):
+        if hasattr(p,"give_mass"):
+            op_arr = cals_in_cd(p)      ## get the time cycles
+        if hasattr(p,"give_time"):
+            op_arr = [{'t_fill': 10, 't_settle': 13, 't_empty': 14}, {'t_fill': 25, 't_settle': 28, 't_empty': 30}, {'t_fill': 35, 't_settle': 38, 't_empty': 40}]
```
can you store this information in the params file instead?

benjym commented on void_migration/void_migration.py:
```diff
@@ -581,15 +820,27 @@ def time_march(p):
 
         if t % p.save_inc == 0:
-            plotter.make_saves(x, y, s, u, v, c, T, p, t)
+            if hasattr(p,"charge_discharge"):
```
can you move this logic into the make_saves function instead to keep the time_march function as clean as possible?

benjym commented on void_migration/void_migration.py:
```diff
@@ -566,6 +791,20 @@ def time_march(p):
         if hasattr(p, "temperature"):
             T = thermal.update_temperature(s, T, p)  # delete the particles at the bottom of the hopper
 
+        if hasattr(p,"charge_discharge"):
```
can you pull this out into a separate function?

benjym commented on void_migration/void_migration.py:
```diff
@@ -581,15 +820,27 @@ def time_march(p):
         # nu_time_x[t] = np.mean(1 - np.mean(np.isnan(s), axis=2), axis=1)
+
         t += 1
+    if hasattr(p,"charge_discharge"):
```
same here, can you move this into the make_saves function?

benjym reviewed (changes requested):
looks much better. just a few small changes and we can merge

Ashivu committed:
further cleaning for charge discharge cycles

further cleaning of charge discharge cycle in void_migration. added the default time for cycles of charge discharge in defaults.json and in params. some changes in plotter.py

benjym committed:
Update defaults.json5

benjym committed:
Update params.py
